compiler-rt/src/crt0/memory/cmpxchg.cairo

```diff
@@ -0,0 +1,247 @@
+use super::super::allocator::{AllocatorState, Address};
+use super::{load::load, store::store};
+use core::num::traits::{BitSize, OverflowingMul};
+use core::traits::{BitAnd, BitOr};
+use super::get_allocator;
+use crate::integer::{u24::u24, u40::u40, u48::u48};
+
+/// The ‘cmpxchg’ instruction is used to atomically modify memory. It loads a value in memory
+/// and compares it to a `expected_value` value. If they are equal, it tries to store `new_value`
+/// into the memory. The value from the memory is returned along with a flag indicating whether the
+/// store did (true) or did not (false) happen.
+///
+/// This `cmpxchg` implementation has strong semantics. That is, if `expected_value` is equal to the
+/// value in memory, `new_value` will always be stored into the memory.
+///
+/// `offset` must be non-negative.
+///
+/// This is the implementation of the `__llvm_cmpxchg_*` polyfills logic. It exists so that it can
+/// be unit-tested with a mock allocator. The polyfills are wrappers over this function and they get
+/// the actual allocator with the getter generated by Hieratika.
+fn cmpxchg<
+    T,
+    +BitAnd<T>,
+    +BitOr<T>,
+    +BitSize<T>,
+    +Copy<T>,
+    +Div<T>,
+    +Drop<T>,
+    +Into<u8, T>,
+    +OverflowingMul<T>,
+    +PartialEq<T>,
+    +TryInto<T, u8>,
+>(
+    ref allocator: AllocatorState, address: Address, offset: i64, expected_value: T, new_value: T,
+) -> (T, bool) {
+    let existing_value = load(ref allocator, address, offset);
+    let did_exchange_happen = if existing_value == expected_value {
+        store(ref allocator, new_value, address, offset);
+        true
+    } else {
+        false
+    };
+    return (existing_value, did_exchange_happen);
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::crt0::allocator::{Allocator, AllocatorOps};
+    use super::load;
+    use core::fmt::Debug;
+
+    /// Prepare allocator for the test suite.
+    ///
+    /// Instantiate the allocator, allocate a memory region and store a predefined array of bytes
+    /// at the beginning of the allocated region. The test cases will load bytes from that region.
+    fn get_allocator() -> Box<AllocatorState> {
+        let mut allocator = Allocator::new();
+        let address1 = allocator.allocate(32);
+        #[cairofmt::skip]
+        let array1 = array![
+            0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f,
```

Comment:
I tend to suggest avoiding zero bytes for things like this. Maybe set the first byte to something other than `0x00`?